src/EditorFeatures/Core.Wpf/InheritanceMargin/InheritanceChainMarginTaggerProvider.cs

```diff
@@ -0,0 +1,132 @@
+                subjectBuffer,
+                TaggerDelay.OnIdle,
+                ThreadingContext,
+                AsyncListener,
+                TaggerEventSources.OnWorkspaceChanged(subjectBuffer, TaggerDelay.OnIdle, AsyncListener),
+                TaggerEventSources.OnViewSpanChanged(ThreadingContext, textViewOpt, TaggerDelay.OnIdle, TaggerDelay.OnIdle),
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer, TaggerDelay.OnIdle),
+                TaggerEventSources.OnOptionChanged(subjectBuffer, FeatureOnOffOptions.ShowInheritanceMargin, TaggerDelay.OnIdle));
+
+        protected override IEnumerable<SnapshotSpan> GetSpansToTag(ITextView textView, ITextBuffer subjectBuffer)
+        {
+            this.AssertIsForeground();
+
+            var visibleSpan = textView.GetVisibleLinesSpan(subjectBuffer, extraLines: 100);
+            if (visibleSpan == null)
+            {
+                return base.GetSpansToTag(textView, subjectBuffer);
+            }
+
+            return SpecializedCollections.SingletonEnumerable(visibleSpan.Value);
+        }
+
+        protected override async Task ProduceTagsAsync(
+            TaggerContext<InheritanceMarginTag> context,
+            DocumentSnapshotSpan spanToTag,
+            int? caretPosition)
+        {
+            var document = spanToTag.Document;
+            if (document == null)
+            {
+                return;
+            }
+
+            var cancellationToken = context.CancellationToken;
+
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var featureEnabled = options.GetOption(FeatureOnOffOptions.ShowInheritanceMargin);
+            if (!featureEnabled)
+            {
+                return;
+            }
+
+            var inheritanceMarginInfoService = document.WithFrozenPartialSemantics(cancellationToken).GetLanguageService<IInheritanceMarginService>();
+            if (inheritanceMarginInfoService == null)
+            {
+                return;
+            }
+
+            var inheritanceMemberItems = await inheritanceMarginInfoService.GetInheritanceInfoAsync(
+                    document,
+                    spanToTag.SnapshotSpan.Span.ToTextSpan(),
+                    cancellationToken).ConfigureAwait(false);
+
+            if (inheritanceMemberItems.IsEmpty)
+            {
+                return;
+            }
+
+            // One line might have multiple members to show, so group them.
+            // For example:
+            // interface IBar { void Foo1(); void Foo2(); }
+            // class Bar : IBar { void Foo1() { } void Foo2() { } }
+            var lineToMembers = inheritanceMemberItems
+                .GroupBy(item => item.LineNumber)
+                .ToImmutableDictionary(
+                    keySelector: grouping => grouping.Key,
+                    elementSelector: grouping => grouping.SelectAsArray(g => g));
```

Comment:
i don't think you need the ToImmutableDictionary at all.  you can just process each group from GroupBy in the foreachloop below.